### ISSUE TITLE
Guide for full finetuning Flex ?

### ISSUE THREAD

**Issue by BenDes21** [tdrussell/diffusion-pipe#89]
Hi there, they're is some infos for train with Hunyuan but it's would be nice to have a mini guide of few steps of how full finetuning Flex and even create lora's on flux, I don't really know what to do except modify the flux dev config file. Thanks you

**Comment by tdrussell** [comment]
For full fine tuning flux, the short answer is you just remove the [adapter] part of the config, and it will full fine tune.

The limiting factor is that you need a lot of VRAM now, and if you are able to fine tune it depends on a combination of your setup and any tricks you do in the configuration to try to save memory.

You'll likely need multiple GPUs, and set pipeline_stages=n_gpus to divide the model across GPUs. And you can try to use certain optimizers and settings to save memory, for example:

```
[optimizer]
type = 'adamw8bitkahan'
lr = 2e-6
betas = [0.9, 0.99]
weight_decay = 0.01
eps = 1e-8
gradient_release = true
```

I haven't tested this in a while, but it uses a custom AdamW8bit optimizer I added that also has kahan summation (required for bf16 training). And additionally turns on a very experimental gradient release setting that frees grad memory and immediately updates the optimizer state during backprop to save memory. With these settings, on 4x4090 with pipeline_stages=4 I was able to full fine tune flux on 512 resolution images.

**Comment by BenDes21** [comment]
> For full fine tuning flux, the short answer is you just remove the [adapter] part of the config, and it will full fine tune.
> 
> The limiting factor is that you need a lot of VRAM now, and if you are able to fine tune it depends on a combination of your setup and any tricks you do in the configuration to try to save memory.
> 
> You'll likely need multiple GPUs, and set pipeline_stages=n_gpus to divide the model across GPUs. And you can try to use certain optimizers and settings to save memory, for example:
> 
> ```
> [optimizer]
> type = 'adamw8bitkahan'
> lr = 2e-6
> betas = [0.9, 0.99]
> weight_decay = 0.01
> eps = 1e-8
> gradient_release = true
> ```
> 
> I haven't tested this in a while, but it uses a custom AdamW8bit optimizer I added that also has kahan summation (required for bf16 training). And additionally turns on a very experimental gradient release setting that frees grad memory and immediately updates the optimizer state during backprop to save memory. With these settings, on 4x4090 with pipeline_stages=4 I was able to full fine tune flux on 512 resolution images.

Hi there! thanks for your answer, about the dataset, is the dataset.toml in the folder example to modify also  ? Thanks

**Comment by tdrussell** [comment]
Yes you can use the example as reference, but modify it according to your desired resolution(s), aspect ratio buckets, num repeats etc.
